### ISSUE TITLE
Deep dive of working of guarantors in Agency Cloud

### ISSUE THREAD

**Issue by cbryanreapit** [reapit/foundations#3829]
**Background context**
To unblock https://github.com/reapit/foundations/issues/3312, we should deep dive this functionality so that we can best know how to apply the proposed new API into the platform.

**Specification**
- Understand data source(s)
- Understand how/if data is shifted around during tenancy phases - we need to make sure that we fully understand any rules around what can and cannot be updated if a tenancy is current
- Propose solution(s) to how guarantor information is presented
- May require broader tenancy changes - ideally not - but everything should be on the table

**Comment by HollyJoyPhillips** [comment]
AC/BA requirement review requested

**Comment by plittlewood-rpt** [comment]
Originally thought we could use existing relationships endpoints, however we need to support additional fields (status and type at the moment) so those endpoints aren't suited. We could possibly expose the relationships from these endpoints though as a nice to have, but that's probably for later

TASK 1
-- Add new applicants API POST /guarantors endpoint which uses a similar mechanism to the relationships endpoint but supports the additional metadata that guarantors can have set (just a status at time of writing). It should not be possible to attach a guarantor to a sales applicant

TASK 3
-- Add new applicants API GET /guarantors and GET /guarantors/{id} which expose a lightweight contact model (with the additional fields) as per the related collection (service already has the Contact entity, but will likely need the Extra entity adding)

TASK 2
-- Add new applicants API DELETE /guarantors/{id} endpoint which users a similar mechanism to the relationships endpoint

TASK 3
-- Add new applicants API PATCH /guarantors/{id} endpoint which allows only the status of the referee to be updated

TASK 4
-- Add new applicants API POST /references endpoint which uses a similar mechanism to the relationships endpoint but supports the additional metadata that guarantors can have set (just type and status at time of writing). It should not be possible to attach a reference to a sales applicant

TASK 5
-- Add new applicants API DELETE /references/{id} endpoint which users a similar mechanism to the relationships endpoint

TASK 6
-- Add new applicants API PATCH /references/{id} endpoint which allows only the status of the referee to be updated

TASK 8
-- Add new applicants API GET /references and GET /references/{id} which expose a lightweight contact model (with the additional fields) as per the related collection (service already has the Contact entity, but will likely need the Extra entity adding)

TASKS 9-16
-- Repeat the above for the tenancy endpoint with the following rules
  - It should not be possible to use the POST/DELETE/PATCH endpoints for references or guarantors once a tenany is current or further along in it's lifetime


**General validations/behaviours**
- Guarantor has extra.type value of LG
- Reference has extra.type value of LR
- Reference Type must be valid enum option
- Reference Status must be valid enum option
- Guarantor Type must be valid enum option
- Guarantor Status must be valid enum option
- It should not be possible to add the same referee or guarantor multiple times to the same applicant or tenancy (this will need to be validated at the domain level)
- It should not be possible to use the POST/DELETE/PATCH endpoints for references or guarantors once a tenany is current or further along in it's lifetime
